lib/core/topologies/shared.js

```diff
@@ -437,6 +439,17 @@ const isRetryableWritesSupported = function(topology) {
   return true;
 };
 
+function handleMMAPError(err) {
+  if (err.code === MMAPv1_RETRY_WRITES_ERROR_CODE && err.errmsg.includes('Transaction numbers')) {
+    // According to the retryable writes spec, we must replace the error message in this case.
+    // We need to replace err.message so the thrown message is correct and we need to replace err.errmsg to meet the spec requirement.
+    err.message =
+      'This MongoDB deployment does not support retryable writes. Please add retryWrites=false to your connection string.';
+    err.errmsg = err.message;
+    throw err;
```

Comment:
I don't think we should be throwing this error. We should be passing it along to the callback.

Reply:
Could we make this function perhaps do something like:

```js
const MMAPv1_RETRY_WRITES_ERROR_MESSAGE = 'This MongoDB deployment does not support retryable writes. Please add retryWrites=false to your connection string.';
function getMMAPError(err) {
  if (err.code !== MMAPv1_RETRY_WRITES_ERROR_CODE || !err.errmsg.includes('Transaction numbers')) {
    return err;
  }
  const newErr = new MongoError({
    message: MMAPv1_RETRY_WRITES_ERROR_MESSAGE',
    errmsg: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
    originalError: err
  });
  return newErr;
}
```

and then we can use it like

```js
if (!isRetryableError(err) || !willRetryWrite) {
  err = getMMAPError(err);
  return callback(err);
}
```